### ISSUE TITLE
"unnecessary negation" (if-not-else) is a bit annoying

### ISSUE THREAD

**Issue by birkenfeld** [Manishearth/rust-clippy#859]
Suppose the following case:
```
        if some_error_condition {
            Err(some_error)
        } else if some_other_error_condition {
            Err(some_other_error)
        } else {
            Ok(result)
        }
```
If `some_other_error_condition` happens to be `a != b`, the lint will trigger. However, I don't want to move the Ok case inbetween the Err cases.

**Comment by oli-obk** [comment]
We could treat if elseif else chains as ordered and not lint.

**Comment by llogiq** [comment]
So I see two possibilities:

* make the lint `allow` by default. It does generate a lot of reports and is probably too opinionated
* have the lint bail on encountering `else if`s

Possibly both.

**Comment by mcarton** [comment]
> make the lint \`allow` by default. It does generate a lot of reports and is probably too opinionated

:+1: I’ve found that lint to be rather annoying myself.

**Comment by GrantGryczan** [comment]
For anyone stumbling upon this in the future by browsing related issues to lints (as I am), this is how I personally would write this example (and I wouldn't except it from the lint):

```rust
        if some_error_condition {
            return Err(some_error);
        }

        if some_other_error_condition {
            return Err(some_other_error);
        }

        Ok(result)
```

Independently of whether I use this lint, I usually use early returns for conditions that branch away from the code's happy path (AKA guard clauses), as it helps reduce my mental overhead when reading a function's control flow. This is just my preference, but figured it might be worth noting.
